File: frontend/src/types/RoutePlaner.ts
import { DeliveryRoute } from './DeliveryRoute';
import { Node } from './Node';
import { Neighborhood } from './Neighborhood';

export class RoutePlaner {
  routes: DeliveryRoute[];
  m: number;
  start: Node;
  maximumCourierDistance: number;

  constructor(
    courierCount: number,
    magazine: Node,
    routes?: DeliveryRoute[],
    maximumCourierDistance = 100
  ) {
    if (!routes) {
      this.routes = Array.from({ length: courierCount }, () =>
        new DeliveryRoute(magazine, maximumCourierDistance)
      );
    } else {
      this.routes = routes.map((route) => route.copy());
    }
    this.maximumCourierDistance = maximumCourierDistance;
    this.m = courierCount;
    this.start = magazine;
  }

  copy(): RoutePlaner {
    return new RoutePlaner(
      this.m,
      this.start,
      this.routes,
      this.maximumCourierDistance
    );
  }

  performNeighborhoodOperation(): RoutePlaner {
    const i = Math.floor(Math.random() * this.m);
    let j = Math.floor(Math.random() * this.m);
    while (j === i) {
      j = Math.floor(Math.random() * this.m);
    }
    const newSol = this.copy();
    Neighborhood.performRandomNeighborhood(newSol.routes[i], newSol.routes[j]);
    return newSol;
  }

  createNewSolution(): RoutePlaner {
    const newSol = new RoutePlaner(
      this.m,
      this.start,
      undefined,
      this.maximumCourierDistance
    );
    const nodes: Node[] = this.routes.flatMap((route) =>
      route.route.map((node) => new Node(node.x, node.y))
    );
    nodes.sort(() => Math.random() - 0.5);
    for (const node of nodes) {
      const i = Math.floor(Math.random() * this.m);
      newSol.routes[i].append(node);
    }
    return newSol;
  }

  fitness(): number {
    return this.routes.reduce((sum, route) => sum + route.cost(), 0);
  }

  distance(): number {
    return this.routes.reduce((sum, route) => sum + route.distance(), 0);
  }

  get(
    key: number | { start?: number; stop?: number; step?: number }
  ): DeliveryRoute | DeliveryRoute[] {
    if (typeof key === 'object') {
      const { start = 0, stop = this.routes.length, step = 1 } = key;
      return this.routes.slice(start, stop).filter((_, index) => index % step === 0);
    }
    return this.routes[key];
  }

  set(
    key: number | { start?: number; stop?: number; step?: number },
    value: DeliveryRoute | DeliveryRoute[]
  ): void {
    if (typeof key === 'object') {
      const { start = 0, stop = this.routes.length } = key;
      this.routes.splice(start, stop - start, ...(value as DeliveryRoute[]));
    } else {
      this.routes[key] = value as DeliveryRoute;
    }
  }

  static createRandomSolution(
    courierCount: number,
    magazine: Node,
    nodes: Node[],
    maximumCourierDistance = 100
  ): RoutePlaner {
    const newSol = new RoutePlaner(courierCount, magazine, undefined, maximumCourierDistance);
    nodes = nodes
      .map((node) => node.copy())
      .sort(() => Math.random() - 0.5);
  
    // Distribute at least one node to each courier
    for (let i = 0; i < courierCount; i++) {
      if (i < nodes.length) {
        newSol.routes[i].append(nodes[i]);
      }
    }
  
    // Distribute the remaining nodes randomly
    for (let i = courierCount; i < nodes.length; i++) {
      const j = Math.floor(Math.random() * courierCount);
      newSol.routes[j].append(nodes[i]);
    }
  
    return newSol;
  }  

  [Symbol.iterator](): Iterator<DeliveryRoute> {
    return this.routes[Symbol.iterator]();
  }
}

File: python - prototype/src/RoutePlaner.py
try:
  from src.DeliveryRoute import DeliveryRoute
  from src.Node import Node
  from src.Neighborhood import Neighborhood
except:
  from DeliveryRoute import DeliveryRoute
  from Node import Node
  from Neighborhood import Neighborhood

import numpy as np
import random
from typing import List, Union
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd


class RoutePlaner:
  def __init__(self, courier_count: int, magazine: Node, routes: Union[None, List[DeliveryRoute]] = None):
    if routes is None:
      self.routes = [DeliveryRoute(magazine) for _ in range(courier_count)]
    else:
      self.routes = [route.copy() for route in routes]
    
    self.m = courier_count
    self.start = magazine

  def copy(self):
    return RoutePlaner(self.m, self.start, self.routes)
  
  
  def perform_neighborhood_operation(self) -> "RoutePlaner":
    # return copy with applied neighborhood operation
    i = np.random.randint(0, self.m)
    j = np.random.randint(0, self.m)
    # in case if there will be 2 indices found
    while j == i:
      j = np.random.randint(0, self.m)
    new_sol = self.copy()
    Neighborhood.perform_random_neighborhood(new_sol[i], new_sol[j])
    return new_sol
  
  def create_new_solution(self) -> "RoutePlaner":
    new_sol = RoutePlaner(self.m, self.start)
    
    # need to copy nodes, to make sure, they are not reference and everything won't crash
    nodes = [Node(node.x, node.y) for delivery_route in self for node in delivery_route]
    random.shuffle(nodes)
    
    for node in nodes:
      i = np.random.randint(0, self.m)
      new_sol[i].append(node)

    return new_sol
    
  def fitness(self):
    return np.sum([route.cost() for route in self.routes])
  
  def __getitem__(self, key):
    if isinstance(key, slice):
      return self.routes[key.start:key.stop:key.step]
    return self.routes[key]

  @staticmethod
  def create_random_solution(courier_count, magazine, nodes : List[Node]):
    new_sol = RoutePlaner(courier_count, magazine)
    
    # need to copy nodes
    random.shuffle([Node(node.x, node.y) for node in nodes])
    
    for node in nodes:
      i = np.random.randint(0, courier_count)
      new_sol[i].append(node)
    return new_sol
  
  def __setitem__(self, key, value : DeliveryRoute):
    if isinstance(value, DeliveryRoute):
      if isinstance(key, slice):
        self.routes[key.start:key.stop:key.step] = value
      else:
        self.routes[key] = value
  
  def __iter__(self):
    return iter(self.routes)
  
  def to_dataframe(self):
    df = pd.DataFrame({'index' : [], 'x': [], 'y': []})
    for i in range(self.m):
      route_df = pd.DataFrame({'index' : [i for _ in range(self[i].size)], 'x': [node.x for node in self[i]], 'y': [node.y for node in self[i]]})
      df = pd.concat([df, route_df])
    return df
  
  def plot_solution(self):  
    paletteName = 'viridis'  
    df = self.to_dataframe()  
    plt.figure(figsize=(10, 6))  
    palette = sns.color_palette(paletteName, self.m)  
      
    sns.scatterplot(data=df, x='x', y='y', hue='index', palette=paletteName, s=100)  
  
    for i in range(self.m):  
      route = self[i]  
      color = palette[i % len(palette)]  
      plt.plot([self.start.x, route[0].x], [self.start.y, route[0].y], color=color, linewidth=2)  
      for j in range(route.size - 1):  
        plt.plot(  
          [route[j].x, route[j + 1].x],  
          [route[j].y, route[j + 1].y],  
          color=color,  
          linewidth=2  
        )  
      plt.plot(  
        [route[-1].x, self.start.x],  
        [route[-1].y, self.start.y],  
        color=color,  
        linewidth=2  
        )  
      
    plt.plot(self.start.x, self.start.y, 'ko', markersize=10)  
  
    # Create legend  
    handles = [plt.Line2D([0], [0], color=palette[i], lw=2, label=f'Route {i+1}') for i in range(self.m)]  
    handles.append(plt.Line2D([0], [0], color='black', marker='o', markersize=5, label='Start'))  # Legend for start point  
    plt.legend(handles=handles, title="Routes")  
  
    plt.title(f"Solution, cost={self.fitness():.2f}")  
    plt.xlabel("X")  
    plt.ylabel("Y")  
    plt.grid(True)  
    plt.show()  